### PR TITLE
Improve the storage writing

### PR DESCRIPTION
**Pull request by MathieuDutSik**
There was more asynchronicity that could be eliminated from the storage writing by using the batches.

### CODE REVIEW

File: linera-storage/src/lib.rs [hunk 1]
```diff
@@ -78,12 +78,29 @@ pub trait Store: Sized {
     /// Writes several values
     async fn write_values(&self, values: &[HashedValue]) -> Result<(), ViewError>;
 
+    /// Writes several values in a batch
+    fn write_values_batch(
```
Comment (afck) [inline linera-storage/src/lib.rs:82]: Maybe we should avoid the word "write" for adding things to a batch, if they are not actually being written to storage at that point.

**Review by afck — APPROVED** [review]

File: linera-core/src/worker.rs [hunk 2]
```diff
@@ -525,8 +526,11 @@ where
         for value in blobs {
             self.cache_recent_value(value.clone());
         }
-        self.storage.write_values(blobs).await?;
-        self.storage.write_certificate(&certificate).await?;
+        let mut batch = Batch::new();
+        self.storage.write_values_batch(blobs, &mut batch)?;
+        self.storage
+            .write_certificate_batch(&certificate, &mut batch)?;
+        self.storage.write_batch(batch).await?;
```
[the remark this answers is not in the document]
Comment (MathieuDutSik) [reply linera-core/src/worker.rs]: Ok, I used two parallel queries for this. For the bunch of certificates, I put a `write_certificates` function that uses batches, but only internally.

File: linera-storage/src/lib.rs [hunk 3]
```diff
@@ -84,6 +84,9 @@ pub trait Store: Sized {
     /// Writes the given certificate.
     async fn write_certificate(&self, certificate: &Certificate) -> Result<(), ViewError>;
 
+    /// Writes a vector of certficates.
```
Comment (ma2bd) [inline linera-storage/src/lib.rs:87]: typo: certficates

File: linera-storage/src/lib.rs [hunk 4]
```diff
@@ -321,20 +316,56 @@ where
...
+    }
+}
+
+impl<CL> DbStoreClient<CL>
+where
+    CL: KeyValueStoreClient + Clone + Send + Sync + 'static,
+    ViewError: From<<CL as KeyValueStoreClient>::Error>,
+    <CL as KeyValueStoreClient>::Error: From<bcs::Error> + Send + Sync + serde::ser::StdError,
+{
+    fn write_value_batch(&self, value: &HashedValue, batch: &mut Batch) -> Result<(), ViewError> {
```
Comment (ma2bd) [inline linera-storage/src/lib.rs:343]: These new `write_xxx_batch` look like variants of `write_batch` which is a bit confusing.
Maybe `add_xxx_to_batch` ?

File: linera-storage/src/lib.rs (hunk 1 above)
Comment (ma2bd) [reply linera-storage/src/lib.rs]: yes, exactly!

File: linera-core/src/worker.rs [hunk 5]
```diff
@@ -525,8 +525,12 @@ where
         for value in blobs {
             self.cache_recent_value(value.clone());
         }
-        self.storage.write_values(blobs).await?;
-        self.storage.write_certificate(&certificate).await?;
+        let (result_blob, result_certificate) = tokio::join!(
```
Comment (ma2bd) [inline linera-core/src/worker.rs:528]: We may want to start using non-tokio equivalent for this kind of stuff.

**Review by ma2bd — COMMENTED** [review]
What happens if we try to write a lot of certificates at once?

**Comment by MathieuDutSik** [comment]
> What happens if we try to write a lot of certificates at once?

It should happen in the same process if we write a lot of data. There is nothing specific about certificates here.
In the case of DynamoDB, it would do some splitting into data blocks as done with journaling. In the case of RocksDB I do not think we have a mechanism for handling. In the case of memory, I believe the question is irrelevant.

File: linera-storage/src/lib.rs (hunk 3 above)
Comment (MathieuDutSik) [reply linera-storage/src/lib.rs:87]: Ok, done.

File: linera-core/src/worker.rs (hunk 5 above)
Comment (MathieuDutSik) [reply linera-core/src/worker.rs:528]: Ok, if you expain what you would like, I could do it.

Comment (ma2bd) [reply linera-core/src/worker.rs:528]: This one https://doc.rust-lang.org/std/future/macro.join.html but sadly it's still unstable.

**Review by ma2bd — APPROVED** [review]
LGTM

**Pull request merged**